Libmesh exception handling

roystgnr opened the pull request:
Refs #31856 - this is the "medium-term" step need to make MOOSE compatible with future fixes at the libMesh level.

MOOSE will now catch std::exception and check what() for keywords like Jacobian and singular to identify recoverable exceptions from degenerate maps regardless of libMesh exception type.

While I was at it I noticed a `MeshDiagnosticsGenerator` exception-catching issue for which @GiudGiud might want to review my fix, as well as a couple places where we were using a `try`/`catch` as a bit of a hack where an `if()` would do.

roystgnr committed:
Catch Jacobian exceptions as non-LogicError

This is needed for forward compatibility of our test suite with upcoming
libMesh exception changes.

Refs #31856

roystgnr committed:
Flexible expect_err in bad-Jacobian tests

An error message calling 0 a negative number should have been noticed as
a bug to fix, not a feature to insist upon!

Refs #0 - see https://github.com/libMesh/libmesh/pull/4286

roystgnr committed:
Prefer if alone over try+if+throw+catch

roystgnr committed:
Fix negative Jacobian/volume diagnostics

The LogicError->exception code here adds forward compatibility with
libMesh; while I was at it I noticed that we promised to count
negative-volume elements but we didn't.

roystgnr committed:
Catch+handle only libMesh DegenerateMap exceptions

Remember what a technological advance the MMU was, because certain types
of errors would now throw a signal that would merely crash the program
rather than potentially corrupting everything running on your computer
from that point onward?  That was great.  That was the correct
philosophy.

roystgnr committed:
Singularity LogicError->DegenerateMap fwd compat

Trying to invert a singular matrix is more of a runtime error than a
logic error, and libMesh will be changing the exception type
accordingly.

roystgnr committed:
Rethrow errors with their same type

If we catch a reference to an error and rethrow *that*, it implicitly
casts the error to that reference class, even if it should be of a
subclass, and the error can then not be caught as that subclass.

roystgnr committed:
Regularize degeneracy exception messages a bit

roystgnr committed:
Include <cstring> whereever we might need it

I had already added this a couple places where a compiler complained,
but I don't want to depend on indirect inclusion of a header that none
of our directly-included headers are explicitly guaranteeing.

moosebuild commented:
Job [Test](https://civet.inl.gov/job/3365712/), step Results summary on e19c068 wanted to post the following:

## Framework test summary

Compared against 7323e92 in job [civet.inl.gov/job/3361666](https://civet.inl.gov/job/3361666).

### Removed tests



### Added tests



### Run time changes
## Modules test summary

Compared against 7323e92 in job [civet.inl.gov/job/3361666](https://civet.inl.gov/job/3361666).

### Removed tests



### Added tests



### Run time changes

lindsayad reviewed (changes requested):
I made a change in the last week to `ElementJacobianDamper.C` introducing another catch block for element Jacobian issues. That could also use your clean up to use `strstr` and to not avoid slicing the derived type information from `e`. You might have to rebase to see it

GiudGiud reviewed (approved):
bug fix in counting negative vol elements looks good to me
other changes too

roystgnr committed:
strstr+rethrow properly in ElementJacobianDamper

This catch postdates my original branch and I didn't think to look for
new catch blocks when rebasing; thanks to Alex for catching it.

moosebuild commented:
Job [Test](https://civet.inl.gov/job/3367487/), step Results summary on dfd0dd3 wanted to post the following:

## Framework test summary

Compared against 7323e92 in job [civet.inl.gov/job/3361666](https://civet.inl.gov/job/3361666).

### Removed tests



### Added tests



### Run time changes
## Modules test summary

Compared against 7323e92 in job [civet.inl.gov/job/3361666](https://civet.inl.gov/job/3361666).

### Removed tests



### Added tests



### Run time changes

| Test                                                                 | Base (s)   | Head (s)   | +/-     |
|----------------------------------------------------------------------|------------|------------|---------|
| `phase_field/test:phase_field_contact_angle.contactangle_150`        | 9.23       | 14.78      | +60.06% |
| `stochastic_tools/test:web_server_control.stochastic_control/normal` | 3.19       | 4.89       | +53.29% |

loganharbour commented:
This broke next: https://civet.inl.gov/job/3368192/

roystgnr commented:
I'll see if I can reproduce
